eureka-client/src/main/java/com/netflix/discovery/converters/Converters.java

```diff
@@ -475,40 +472,35 @@ public void marshal(Object source, HierarchicalStreamWriter writer,
         @SuppressWarnings("unchecked")
         public Object unmarshal(HierarchicalStreamReader reader,
                                 UnmarshallingContext context) {
-            DataCenterInfo info = null;
+            MetadataAware info = null; // will also always be a DataCenterInfo implementation
+            Map<String, String> metadataMap = Collections.emptyMap();
+
             while (reader.hasMoreChildren()) {
                 reader.moveDown();
 
                 if (ELEM_NAME.equals(reader.getNodeName())) {
                     final String dataCenterName = reader.getValue();
-                    if (DataCenterInfo.Name.Amazon.name().equalsIgnoreCase(
-                            dataCenterName)) {
+                    if (DataCenterInfo.Name.Amazon.name().equalsIgnoreCase(dataCenterName)) {
                         info = new AmazonInfo();
                     } else {
-                        final DataCenterInfo.Name name =
-                                DataCenterInfo.Name.valueOf(dataCenterName);
-                        info = new DataCenterInfo() {
-
-                            @Override
-                            public Name getName() {
-                                return name;
-                            }
-                        };
+                        info = new MyDataCenterInfo(DataCenterInfo.Name.valueOf(dataCenterName));
```

Comment:
I suppose we would use `MyDataCenterInfo` rather than a custom type.  Because of all that silliness around serializing any non-AmazonInfo with a @class attribute of  `com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo`, we are down to a binary choice: AmazonInfo or not.  At least this way we can hang metadata on our non-AmazonInfo choice.